sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -1658,19 +1658,14 @@ class Analyzer(
       // Third, for every Window Spec, we add a Window operator and set currentChild as the
       // child of it.
       var currentChild = child
-      var i = 0
-      while (i < groupedWindowExpressions.size) {
-        val ((partitionSpec, orderSpec), windowExpressions) = groupedWindowExpressions(i)
+      groupedWindowExpressions.foreach { case ((partitionSpec, orderSpec), windowExpressions) =>
```

Comment:
Well `currentChild` is a `Window`, and that's the only thing whose computation is changed here. How about:

```
      val currentChild =
        groupedWindowExpressions.foldLeft(child) {
          case (last, ((partitionSpec, orderSpec), windowExpressions)) =>
            Window(windowExpressions, partitionSpec, orderSpec, last)
        }
```